std/functional.d

```diff
@@ -716,15 +716,9 @@ Returns:
  */
 template partial(alias fun, alias arg)
 {
-    static if (is(typeof(fun) == delegate) || is(typeof(fun) == function))
-    {
-        import std.traits : ReturnType;
-        ReturnType!fun partial(Parameters!fun[1..$] args2)
-        {
-            return fun(arg, args2);
-        }
-    }
-    else
+    import std.traits : isCallable;
+    // Check whether fun is a user defined type which implements opCall or a template.
+    static if (is(typeof(fun) == struct) || is(typeof(fun) == class) || __traits(isTemplate, fun))
```

Comment:
Looks like I missed this on my first time through. If `fun` is a struct or class, there should also be a check that they are actually callable. In fact, your check for `isCallable!fun` should either be moved up as a static assert at the top of the template, or added as a template guard.

After you make these changes we can get this merged, finally.

Reply:
Sorry for taking so long. I experimented a bit and the submitted solution (check for the presence of opCall) is the only one that passes all existing tests and gives a compile error on types that are not callable. I'm still not 100% satisfied with the check, though, as cases that obviously should not compile (e.g. a class with opCall(int, int) being called with string as a first argument) still compile. However, checking that quickly escalates the complexity, which is probably why partial for functors is implemented with templates in the first place.